rustworkx-core/src/generators/random_graph.rs

```diff
@@ -305,6 +305,133 @@ where
     Ok(graph)
 }
 
+/// Generate a graph from the stochastic block model.
+///
+/// The stochastic block model is a generalization of the G<sub>np</sub> graph model
+/// (see [rustworkx_core::generators::gnp_random_graph] ). The connection probability of
+/// nodes `u` and `v` depends on their block (or community) and is given by
+/// `probabilities[blocks[u]][blocks[v]]`. The number of nodes and the number of blocks
+/// are inferred from `blocks`.
+///
+/// Arguments:
+///
+/// * `blocks` - Block membership (between 0 and B-1) of each node.
+/// * `probabilities` - B x B matrix that contains the connection probability between
+///     nodes of different blocks. Must be symmetric for undirected graphs.
+/// * `loops` - Determines whether the graph can have loops or not.
+/// * `seed` - An optional seed to use for the random number generator.
+/// * `default_node_weight` - A callable that will return the weight to use
+///     for newly created nodes.
+/// * `default_edge_weight` - A callable that will return the weight object
+///     to use for newly created edges.
+///
+/// # Example
+/// ```rust
+/// use rustworkx_core::petgraph;
+/// use rustworkx_core::generators::sbm_random_graph;
+///
+/// let g = sbm_random_graph::<petgraph::graph::DiGraph<(), ()>, (), _, _, ()>(
+///     &vec![1, 0, 1],
+///     &[vec![0., 1.], vec![0., 1.]],
+///     true,
+///     Some(10),
+///     || (),
+///     || (),
+/// )
+/// .unwrap();
+/// assert_eq!(g.node_count(), 3);
+/// assert_eq!(g.edge_count(), 6);
+/// ```
+pub fn sbm_random_graph<G, T, F, H, M>(
+    blocks: &[usize],
+    probabilities: &[Vec<f64>],
```

Comment:
You can import the ndarray crate and have this a 2d array

Reply:
Is rustworkx-core intended to work with ndarray? I feel like I should avoid adding a dependency if possible. I agree that for the Python API it makes sense to directly support a numpy ndarray.

Reply:
Eventually we will have to add ndarray because of some of the centrality functions/Floyd-Warshall. So you can be the first one to add it

Reply:
cc @mtreinish for asking a second opinion

Reply:
I'm fine adding an ndarray dependency on rustworkx core. I think eventually we'll want to start using features and optional dependencies. But for right now I don't see a problem with doing this for core, especially because we're using it in the python side already.